core/App/ffi_operation.cpp

```diff
@@ -0,0 +1,2063 @@
+        {
+            code: int,
+            message: string,
+            result: {}
+        }
+    */
+    char *ffi_initialize()
+    {
+        log_i("Service name:\t\teHSM-KMS service %s", EHSM_VERSION);
+        log_i("Service built:\t\t%s", EHSM_DATE);
+        log_i("Service git_sha:\t\t%s", EHSM_GIT_SHA);
+
+        RetJsonObj retJsonObj;
+        ehsm_status_t ret = EH_OK;
+
+        ret = Initialize();
+        if (ret != EH_OK)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+        }
+        return retJsonObj.toChar();
+    }
+
+    /*
+    destory the enclave
+    */
+    void ffi_finalize()
+    {
+        Finalize();
+    }
+
+    /**
+     * @brief Create key and save the parameters when using the key for encrypt, decrypt, sign and verify
+     *
+     * @param payload : Pass in the key parameter in the form of JSON string
+                {
+                    keyspec : int,
+                    purpose : int,
+                    origin : int,
+                    padding_mode : int,
+                    digest_mode : int
+                }
+     *
+     * @return char*
+     * [string] json string
+        {
+            code: int,
+            message: string,
+            result: {
+                cmk : a base64 string
+            }
+        }
+     */
+    char *ffi_createKey(const char *payload)
+    {
+        RetJsonObj retJsonObj;
+        ehsm_status_t ret = EH_OK;
+        ehsm_keyblob_t *master_key;
+        string cmk_base64;
+        uint8_t *resp = NULL;
+        uint32_t resp_len = 0;
+        if (payload == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception1.");
+            return retJsonObj.toChar();
+        }
+        // parse payload into paramJsonObj
+        JsonObj paramJsonObj;
+        if (!paramJsonObj.parse(payload))
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception2.");
+            return retJsonObj.toChar();
+        }
+
+        master_key = (ehsm_keyblob_t*)malloc(sizeof(ehsm_keyblob_t));
+        if (master_key == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception3.");
+            goto out;
+        }
+        // storage common key properties into metadata of master_key
+        master_key->metadata.keyspec = (ehsm_keyspec_t)paramJsonObj.readData_uint16("keyspec");
+        master_key->metadata.origin = (ehsm_keyorigin_t)paramJsonObj.readData_uint16("origin");
+        master_key->metadata.purpose = (ehsm_keypurpose_t)paramJsonObj.readData_uint16("purpose");
+        master_key->metadata.padding_mode = (ehsm_padding_mode_t)paramJsonObj.readData_uint16("padding_mode");
+        master_key->metadata.digest_mode = (ehsm_digest_mode_t)paramJsonObj.readData_uint16("digest_mode");
```

Comment:
why call it as readData_uint16?

Reply:
and why not put them into a new split function

Reply:
It will be change to uint32 and move to marshal cpp.